File: app/src/main/java/com/pandacorp/taskui/SetTaskActivity.java
package com.pandacorp.taskui;

import android.app.DatePickerDialog;
import android.app.TimePickerDialog;
import android.content.ContentValues;
import android.content.Context;
import android.content.SharedPreferences;
import android.database.sqlite.SQLiteDatabase;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RadioGroup;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.preference.PreferenceManager;

import com.pandacorp.taskui.Adapter.ListItem;
import com.pandacorp.taskui.Notifications.NotificationUtils;
import com.pandacorp.taskui.Widget.WidgetProvider;
import com.pandacorp.taskui.ui.settings.MySettings;

import java.util.Calendar;

public class SetTaskActivity extends AppCompatActivity implements View.OnClickListener {
    private String TAG = "MyLogs";

    private Button set_task_accept_btn;
    private Button set_time_btn;
    private EditText set_task_editText;
    private RadioGroup set_priority_radio_group;

    private DatePickerDialog datePickerDialog;
    private TimePickerDialog timePickerDialog;

    Calendar now = Calendar.getInstance();

    //SQLite database objects.
    private DBHelper dbHelper;
    private SQLiteDatabase database;

    private boolean isTimeSet = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        MySettings mySettings = new MySettings(this);
        mySettings.load();
        setContentView(R.layout.activity_set_task);
        initViews();

    }
    private void initViews(){
        Toolbar toolbar = findViewById(R.id.set_task_toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setHomeButtonEnabled(true);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        set_task_accept_btn = findViewById(R.id.set_task_accept_btn);
        set_task_accept_btn.setOnClickListener(this);
        set_time_btn = findViewById(R.id.set_time_btn);
        set_time_btn.setOnClickListener(this);
        set_task_editText = findViewById(R.id.set_task_editText);

        set_priority_radio_group = findViewById(R.id.set_priority_radio_group);

        initDialogPickers();

        dbHelper = new DBHelper(this);
        database = dbHelper.getWritableDatabase();

    }
    private void initDialogPickers(){
        String pickerDialogThemeSP = PreferenceManager.getDefaultSharedPreferences(this).getString("Themes", MySettings.Theme_Blue);

        int pickerDialogTheme;
        switch (pickerDialogThemeSP){
            case MySettings.Theme_Blue:
                pickerDialogTheme = R.style.BlueTheme_TimePicker;
                break;
            case MySettings.Theme_Dark:
                pickerDialogTheme = R.style.DarkTheme_TimePicker;
                break;
            case MySettings.Theme_Red:
                pickerDialogTheme = R.style.RedTheme_TimePicker;
                break;
            default:
                throw new IllegalStateException("Unexpected value: " + pickerDialogThemeSP);
        }
        DatePickerDialog.OnDateSetListener datePickerDialogListener = (view, year, month, dayOfMonth) -> {
            now.set(Calendar.YEAR, year);
            now.set(Calendar.MONTH, month);
            now.set(Calendar.DAY_OF_MONTH, dayOfMonth);
            timePickerDialog.show();
        };
        TimePickerDialog.OnTimeSetListener timePickerDialogListener = (view, hourOfDay, minute) -> {
            now.set(Calendar.HOUR_OF_DAY, hourOfDay);
            now.set(Calendar.MINUTE, minute);
            now.set(Calendar.SECOND, 0);

            String task_time = String.format("%02d:%02d", now.get(Calendar.HOUR_OF_DAY), now.get(Calendar.MINUTE));
            set_time_btn.setText(task_time);
            isTimeSet = true;
        };
        datePickerDialog = new DatePickerDialog(this,
                pickerDialogTheme,
                datePickerDialogListener,
                now.get(Calendar.YEAR),
                now.get(Calendar.MONTH),
                now.get(Calendar.DAY_OF_MONTH));

        timePickerDialog = new TimePickerDialog(
                this,
                pickerDialogTheme,
                timePickerDialogListener,
                now.get(Calendar.HOUR_OF_DAY),
                now.get(Calendar.MINUTE),
                true
        );
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.set_task_accept_btn:

                if (set_task_editText.getText().length() != 0) {
                    setTask();
                    updateWidgetOnItemAdded();

                }
                break;
            case R.id.set_time_btn:
                datePickerDialog.show();
                break;
        }
    }

    private void setTask() {
        String task_priority;


        String task_text = String.valueOf(set_task_editText.getText());
        //Needed for set time from 10:0 to 10:00
        String task_time = String.format("%02d:%02d", now.get(Calendar.HOUR_OF_DAY), now.get(Calendar.MINUTE));

        switch (set_priority_radio_group.getCheckedRadioButtonId()) {
            case R.id.white_radio_button:
                task_priority = ListItem.white;
                break;
            case R.id.orange_radio_button:
                task_priority = ListItem.yellow;
                break;
            case R.id.red_radio_button:
                task_priority = ListItem.red;
                break;
            default:
                throw new IllegalStateException("SetTaskActivity.settask: Unexpected value: " + set_priority_radio_group.getCheckedRadioButtonId());
        }

        Log.d(TAG, "onClick: task_time = " + task_time);
        Log.d("MyLogs", "onClick: task_text = " + task_text);
        Log.d(TAG, "onClick: task_prioriy = " + task_time);

        ContentValues contentValues = new ContentValues();

        contentValues.put(DBHelper.KEY_TASK_TEXT, task_text);
        contentValues.put(DBHelper.KEY_TASK_PRIORITY, task_priority);
        if (isTimeSet) {
            contentValues.put(DBHelper.KEY_TASK_TIME, task_time);
            setNotification();

        }

        database.insert(DBHelper.MAIN_TASKS_TABLE_NAME, null, contentValues);

        dbHelper.close();

        setResult(RESULT_OK);

        this.finish();
    }

    private void setNotification() {
        long time = now.getTimeInMillis() - Calendar.getInstance().getTimeInMillis();
        long _currentTime = System.currentTimeMillis();
        long _triggerReminder = _currentTime + time;
        String title = getResources().getString(R.string.show_layout_task);
        String content = set_task_editText.getText().toString();

        int notification_id = now.get(Calendar.HOUR_OF_DAY) + now.get(Calendar.MINUTE) + now.get(Calendar.SECOND);
        SharedPreferences sp = getSharedPreferences("notifications_id", Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sp.edit();
        editor.putInt(set_task_editText.getText().toString(), notification_id);

        editor.apply();

        NotificationUtils _notificationUtils = new NotificationUtils(this);
        _notificationUtils.setNotification(title, content);
        _notificationUtils.setReminder(_triggerReminder);


    }

    private void updateWidgetOnItemAdded() {
        WidgetProvider.Companion.sendRefreshBroadcast(this);

    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home: //Метод обработки нажатия на кнопку home.
                this.finish();
                return super.onOptionsItemSelected(item);
            default:
                return super.onOptionsItemSelected(item);
        }

    }

}

File: app/src/main/java/com/pandacorp/taskui/presentation/ui/SetTaskActivity.kt
package com.pandacorp.taskui.presentation.ui

import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.datepicker.MaterialDatePicker
import com.google.android.material.timepicker.MaterialTimePicker
import com.google.android.material.timepicker.TimeFormat
import com.pandacorp.taskui.R
import com.pandacorp.taskui.databinding.ActivitySetTaskBinding
import com.pandacorp.taskui.domain.models.TaskItem
import com.pandacorp.taskui.domain.usecases.AddTaskUseCase
import com.pandacorp.taskui.presentation.notifications.NotificationUtils
import com.pandacorp.taskui.presentation.ui.widget.WidgetProvider
import com.pandacorp.taskui.presentation.utils.Constants
import com.pandacorp.taskui.presentation.utils.PreferenceHandler
import com.pandacorp.taskui.presentation.utils.Utils
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.text.DateFormat
import java.util.Calendar
import java.util.Locale
import javax.inject.Inject

@AndroidEntryPoint
class SetTaskActivity : AppCompatActivity() {
    companion object {
        const val TAG = "SetTaskActivity"
    }

    @Inject
    lateinit var addTaskUseCase: AddTaskUseCase

    private var _binding: ActivitySetTaskBinding? = null
    private val binding get() = _binding!!

    private val dateFormatter = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.SHORT, Locale.getDefault())
    private var selectedDateCalendar = Calendar.getInstance(Locale.getDefault())
    private var selectedTime: Calendar? = null
    private lateinit var timePickerDialog: MaterialTimePicker
    private lateinit var datePickerDialog: MaterialDatePicker<*>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Utils.setupExceptionHandler()
        PreferenceHandler(this).load()
        _binding = ActivitySetTaskBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setSupportActionBar(binding.setTaskToolbarInclude.toolbar)
        supportActionBar!!.setHomeButtonEnabled(true)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setTitle(R.string.createTask)

        // Restore selected values on the device rotation
        savedInstanceState?.apply {
            val time = getLong(Constants.TIME, 0L)
            if (time != 0L) {
                selectedDateCalendar.timeInMillis = time
                selectedTime!!.timeInMillis = time
                binding.setTimeBtn.text = dateFormatter.format(selectedTime!!.time)
            }
            binding.setTaskEditText.setText(getString(Constants.TITLE, null))
            binding.setPriorityRadioGroup.check(getInt(Constants.PRIORITY, 0))

            // Close the opened dialogs due to the callbacks loosing after the device rotation.
            val dialog = supportFragmentManager.findFragmentByTag("DatePickerDialog")
            if (dialog != null) {
                supportFragmentManager.beginTransaction().remove(dialog).commit()
            }
            val dialog2 = supportFragmentManager.findFragmentByTag("TimePickerDialog")
            if (dialog2 != null) {
                supportFragmentManager.beginTransaction().remove(dialog2).commit()
            }
        }

        initViews()
    }

    private fun initViews() {
        datePickerDialog = MaterialDatePicker.Builder.datePicker().build().apply {
            addOnPositiveButtonClickListener { selectedDate ->
                selectedDateCalendar.timeInMillis = selectedDate
                timePickerDialog.show(supportFragmentManager, "TimePickerDialog")
            }
        }

        timePickerDialog = MaterialTimePicker.Builder().setTimeFormat(TimeFormat.CLOCK_24H)
            .setHour(Calendar.getInstance(Locale.getDefault())[Calendar.HOUR_OF_DAY])
            .setMinute(Calendar.getInstance(Locale.getDefault())[Calendar.MINUTE]).build().apply {
                addOnPositiveButtonClickListener {
                    selectedDateCalendar[Calendar.HOUR_OF_DAY] = timePickerDialog.hour
                    selectedDateCalendar[Calendar.MINUTE] = timePickerDialog.minute
                    selectedDateCalendar[Calendar.SECOND] = 0

                    selectedTime = Calendar.getInstance(Locale.getDefault())
                    selectedTime!!.timeInMillis = selectedDateCalendar.timeInMillis

                    binding.setTimeBtn.text = dateFormatter.format(selectedTime!!.time)
                }
            }
        binding.OKBtn.setOnClickListener {
            if (binding.setTaskEditText.text.isNotEmpty()) {
                setTask()
                WidgetProvider.update(this)
            }
        }
        binding.setTimeBtn.setOnClickListener {
            if (supportFragmentManager.findFragmentByTag("DatePickerDialog") != null) return@setOnClickListener // avoid multi clicks if the dialog is already started
            datePickerDialog.show(supportFragmentManager, "DatePickerDialog")
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putLong(Constants.TIME, selectedTime?.timeInMillis ?: 0L)
        outState.putString(Constants.TITLE, binding.setTaskEditText.text.toString())
        outState.putString(Constants.PRIORITY, binding.setPriorityRadioGroup.checkedRadioButtonId.toString())
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                finish()
                true
            }

            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onDestroy() {
        _binding = null
        super.onDestroy()
    }

    private fun setTask() {
        val text = binding.setTaskEditText.text.toString()
        val priority = when (binding.setPriorityRadioGroup.checkedRadioButtonId) {
            R.id.nullRadioButton -> null
            R.id.whiteRadioButton -> TaskItem.WHITE
            R.id.yellowRadioButton -> TaskItem.YELLOW
            R.id.redRadioButton -> TaskItem.RED
            else -> throw IllegalStateException("Unexpected value: ${binding.setPriorityRadioGroup.checkedRadioButtonId}")
        }
        CoroutineScope(Dispatchers.IO).launch {
            val taskItem = TaskItem(text = text, time = selectedTime?.timeInMillis, priority = priority)
            taskItem.id = addTaskUseCase(taskItem)
            val intent = Intent()
            intent.putExtra(Constants.IntentItem, taskItem)
            setResult(RESULT_OK, intent)
            withContext(Dispatchers.Main) {
                if (taskItem.time != null)
                    NotificationUtils.create(this@SetTaskActivity, taskItem)
                WidgetProvider.update(this@SetTaskActivity)
                finish()
            }
        }
    }
}